crates/re_types_builder/src/codegen/python.rs

```diff
@@ -0,0 +1,994 @@
+            }
+            std::collections::hash_map::Entry::Vacant(entry) => {
+                entry.insert(names.clone());
+            }
+        };
+
+        filepaths.push(filepath.clone());
+        let mut file = std::fs::File::create(&filepath)
+            .with_context(|| format!("{filepath:?}"))
+            .unwrap();
+
+        let mut code = String::new();
+        code.push_str(&format!("# {AUTOGEN_WARNING}\n\n"));
+
+        let manifest = quote_manifest(names);
+        code.push_str(&unindent::unindent(&format!(
+            "
+            from __future__ import annotations
+
+            __all__ = [{manifest}]
+
+            ",
+        )));
+
+        for obj in objs {
+            code.push_str(&obj.code);
+            code.push('\n');
+        }
+        file.write_all(code.as_bytes())
+            .with_context(|| format!("{filepath:?}"))
+            .unwrap();
+    }
+
+    // rerun/{datatypes|components|archetypes}/__init__.py
+    {
+        let path = out_path.join("__init__.py");
+
+        let mut code = String::new();
+
+        let manifest = quote_manifest(mods.iter().flat_map(|(_, names)| names.iter()));
+
+        code.push_str(&format!("# {AUTOGEN_WARNING}\n\n"));
+        code.push_str(&unindent::unindent(&format!(
+            "
+            from __future__ import annotations
+
+            __all__ = [{manifest}]
+
+            # NOTE: we use fully qualified paths to prevent lazy circular imports.
+            ",
+        )));
+
+        for (module, names) in &mods {
+            let names = names.join(", ");
+            code.push_str(&format!("from .{module} import {names}\n"));
+        }
+
+        filepaths.push(path.clone());
+        std::fs::write(&path, code)
+            .with_context(|| format!("{path:?}"))
+            .unwrap();
+    }
+
+    (filepaths, all_names)
+}
+
+// --- Codegen core loop ---
+
+#[derive(Debug, Clone)]
+struct QuotedObject {
+    filepath: PathBuf,
+    name: String,
+    kind: ObjectKind,
+    code: String,
+}
+
+impl QuotedObject {
+    fn from_struct(arrow_registry: &ArrowRegistry, objects: &Objects, obj: &Object) -> Self {
+        assert!(obj.is_struct());
+
+        let Object {
+            filepath,
+            fqname: _,
+            pkg_name: _,
+            name,
+            docs,
+            kind,
+            attrs: _,
+            fields,
+            specifics: _,
+        } = obj;
+
+        let mut code = String::new();
+
+        code.push_str(&quote_module_prelude());
+
+        for clause in obj
+            .fields
+            .iter()
+            .filter_map(quote_import_clauses_from_field)
+        {
+            code.push_str(&clause);
+            code.push('\n');
+        }
+
+        code.push_str(&unindent::unindent(&format!(
+            r#"
+
+            @dataclass
+            class {name}:
+            "#
+        )));
+
+        code.push_str(&indent::indent_all_by(4, quote_doc_from_docs(docs)));
```

Comment:
as mentioned on the call, `indent::indent_all_by(4,` drives me nuts pretty quickly, I'd add some utility for that
(if you don't do it now I might later ;))